repair/repair.cc

```diff
@@ -1959,23 +1985,66 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get0();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_live_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    size_t nr_racks = source_dc_racks.size();
+                    size_t max_replicas_per_rack = (rf + nr_racks - 1) / nr_racks;
```

Comment:
This calculation looks correct. Could you  add a comment how this is calculated?

Reply:
sure.  I'll mention the way network_topology_replication_strategy fills the replicas in round robin across the racks.
and that said, now that I think of that, if different racks have different number of nodes there could be pathological cases where there could be more replicas in a specific rack than that...
For example if a dc has this topology across racks: `[1, 1, 3]` and rf=5, the calculation above will return `2 = (5 + 2 / 3)`, since it assumes that all racks are symmetrical, but in reality all nodes will be populated with replicas.
I'm not sure how realistic it is or how much it matters for the quorum calculation below.
It certainly is no worse than before....

Reply:
See v5